cluster/kubernetes/sshkeyring.go

```diff
@@ -0,0 +1,130 @@
+package kubernetes
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"k8s.io/client-go/1.5/kubernetes/typed/core/v1"
+	"k8s.io/client-go/1.5/pkg/api"
+
+	"github.com/weaveworks/flux/ssh"
+)
+
+const (
+	// The private key file must have these permissions, or ssh will refuse to
+	// use it
+	privateKeyFileMode = os.FileMode(0400)
+)
+
+// sshKeyRingConfig is used to configure the keyring with key generation
+// options and the parameters of its backing kubernetes secret resource.
+// secretVolumeMountPath must be mounted RW for regenerate() to work, and to
+// set the privateKeyFileMode on the identity secret file.
+type sshKeyRingConfig struct {
+	secretAPI             v1.SecretInterface
+	secretVolumeMountPath string // e.g. "/etc/fluxd/ssh"
+	secretDataKey         string // e.g. "identity"
+	keyBits               ssh.OptionalValue
+	keyType               ssh.OptionalValue
+}
+
+type sshKeyRing struct {
+	sync.RWMutex
+	sshKeyRingConfig
+	publicKey      []byte
+	privateKeyPath string
+}
+
+// newSSHKeyRing constructs an sshKeyRing backed by a kubernetes secret
+// resource. The keyring is initialised with the key that was previously stored
+// in the secret (either by regenerate() or an administrator), or a freshly
+// generated key if none was found.
+func newSSHKeyRing(config sshKeyRingConfig) (*sshKeyRing, error) {
```

Comment:
> I was expecting us to use the k8s Go API to access the secrets directly.

I didn't want to read the private key into memory any more than necessary, particularly in a long lived process where we have little control over the memory management (other than mclockall future which is a sledgehammer). This way, under normal operation k8s puts the secret in a tmpfs where it it accessed directly by `ssh`, which is written to handle the private key in memory in a secure fashion.